src/FTThresholdClient.cpp

```diff
@@ -0,0 +1,79 @@
+#include "rewd_controllers/FTThresholdClient.hpp"
+#include <thread>
+
+namespace rewd_controllers {
+
+using SetFTThresholdAction = pr_control_msgs::SetForceTorqueThresholdAction;
+using FTThresholdActionClient
+    = actionlib::SimpleActionClient<SetFTThresholdAction>;
+
+//=============================================================================
+FTThresholdClient::FTThresholdClient(const std::string& controllerThresholdTopic, ros::NodeHandle nodeHandle)
+  : nodeHandle(nodeHandle)
+{
+  ftThresholdActionClient = std::unique_ptr<FTThresholdActionClient>(
+      new FTThresholdActionClient(controllerThresholdTopic));
+  ROS_INFO("Waiting for FT Threshold Action Server to start...");
+  ftThresholdActionClient->waitForServer();
+  ROS_INFO("FT Threshold Action Server started.");
+}
+
+//=============================================================================
+bool FTThresholdClient::trySetThresholdRepeatedly(double forceThreshold, double torqueThreshold)
+{
+  bool setFTSuccessful = false;
+  while (!setFTSuccessful)
+  {
+    setFTSuccessful = trySetThreshold(forceThreshold, torqueThreshold);
+    if (setFTSuccessful)
+      break;
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    if (!ros::ok())
+    {
+      return false;
+    }
+  }
+  return true;
+}
+
+//=============================================================================
+bool FTThresholdClient::trySetThreshold(double forceThreshold, double torqueThreshold, double timeout)
+{
+  try
```

Comment:
Alright.

Reply:
On a second look: If we flip these functions, we can't really give a good error message anymore in the failure case.
Add a `std::string&` parameter or keep the functions as they are?

Reply:
I think `setThresholds` should still throw an error in if there is a failure due to bad values, e.g. negative thresholds. My argument for this is that this _should_ be catastrophic because the user was wrong, and the error shouldn't just be squashed along with timeout errors. (That does mean that there are now two possible places that errors are thrown: `setThresholds` throws if the action server rejects them and `setThresholdsOrThrow` throws if there's a timeout.) Does that sound okay?

I think this function should basically be identical to the current implementation of `setThresholdOrThrow`, except for the final throw as a result of the timeout. Instead, we can just return `finished_before_timeout`.